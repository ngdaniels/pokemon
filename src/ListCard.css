.card {
    border-radius: 1rem;
    border: 2px solid grey;
    margin: 1rem;
    padding: 1rem;
    width: calc(100% - 2rem);
}
            
.pokemon-picture {
    float: right;
    height: 100%;
}

.release {
    display: flex;
    justify-content: flex-end;
}

.release-button {
    background-color: tomato;
    font-weight: bold;
    color: white;
    border: none;
    border-radius: 2rem;
    padding: 0.5rem 1rem;
    text-align: right;
    width: 100%;
    height: 1rem;;
    margin: 1px;
}

.confirm-release {
    background-color: turquoise !important;
    text-align: center !important;
    width: 1.5rem !important;
}

.confirmation-text {
    white-space: normal;
    overflow-wrap: break-word;
    max-width: calc(100% - 3rem);
    margin-right: auto;
}

