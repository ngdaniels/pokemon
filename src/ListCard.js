import React, { Component } from "react";
import './ListCard.css';
import $ from 'jquery';
import {
    Redirect
} from "react-router-dom";

class ListCard extends Component {
    constructor(props) {
        super(props);
        this.state = {
            releasing: false,
            showConfirmationText: false
        };
    }

    goToDetailPage = () => {
        this.setState({
            goToDetail: true
        });
    };

    toggleReleaseConfirmation = async () => {
        let { pokemon } = this.props;

        this.setState({
            releasing: !this.state.releasing
        });
        if (this.state.releasing) {
            this.setState({
                showConfirmationText: false
            });
            await $(`#release-${pokemon.name}-${pokemon.nickname}`).animate({ width: "100%" }).promise();
        }
        else {
            await $(`#release-${pokemon.name}-${pokemon.nickname}`).animate({ width: "1.5rem" }).promise();
            this.setState({
                showConfirmationText: true
            });
        }
    };

    release = () => {
        this.setState({
            deleted: true
        });

        setTimeout(() => {
            this.props.release(this.props.pokemon.nickname);
        }, 300);
        
    }

    render() {
        let { pokemon, page } = this.props;

        // Get Pokemon Number
        let pokemonNumber = pokemon.id;

        let diplayedPokemonNumber = String(pokemonNumber);
        while (diplayedPokemonNumber.length < 3) {
            diplayedPokemonNumber = `0${diplayedPokemonNumber}`;
        }
        diplayedPokemonNumber = `#${diplayedPokemonNumber}`

        let topColumn, botColumn, release;

        if (page === 'list') {
            topColumn = (
                <span className="bold">{diplayedPokemonNumber}</span>
            );
            botColumn = (
                <span className="bold">{pokemon.name.charAt(0).toUpperCase() + pokemon.name.slice(1)}</span>
            );
            release = null;
        }
        else if (page === 'collection') {
            topColumn = (
                <span className="bold">{pokemon.nickname}</span>
            );
            botColumn = (
                <div className="bold">
                    {diplayedPokemonNumber}<br></br>
                    {pokemon.name.charAt(0).toUpperCase() + pokemon.name.slice(1)}
                </div>
            );

            let releaseConfirmationText = this.state.showConfirmationText ? <div className="confirmation-text">Are you sure to release {pokemon.nickname}?</div> : '';
            let releaseConfirmationButton = this.state.releasing ? <div className="card-button confirm-release" onClick={this.release}>✔</div> : '';
            let releaseButtonText = this.state.releasing ? "✘" : "Release";

            release = (
                <tfoot>
                    <tr>
                        <td colSpan="2" >
                            <div className="flex-end">
                                {releaseConfirmationText}
                                {releaseConfirmationButton}

                                <div className="card-button button-release" id={`release-${pokemon.name}-${pokemon.nickname}`} onClick={this.toggleReleaseConfirmation}>
                                    {releaseButtonText}
                                </div>
                            </div>
                        </td>
                    </tr>
                </tfoot>
            );
        }

        let cardTable = (
            <tbody onClick={this.goToDetailPage}>
                <tr>
                    <td className="elipsis" style={{ width: "50%" }}>
                        {topColumn}
                    </td>
                    <td rowSpan="2">
                        <img className="pokemon-picture" src={pokemon.image} alt={pokemon.name}></img>
                    </td>
                </tr>
                <tr>
                    <td className="elipsis">
                        {botColumn}
                    </td>
                </tr>
            </tbody>
        );

        if (this.state.goToDetail) {
            return (
                <Redirect to={`detail/${pokemon.name}`} />
            )
        }

        if (page === "collection" || page === "list") {
            return (
                <table className="card" style={ {transform: `scaleY(${this.state.deleted ? 0 : 1})` } }>
                    {cardTable}
                    {release}
                </table>
            );
        }
        else {
            return (
                ""
            )
        }
    }
}

export default ListCard;