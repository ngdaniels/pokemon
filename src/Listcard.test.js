import { render } from '@testing-library/react';
import ListCard from './ListCard';

describe('List Card', () => {
	test('List Card Content for Pokemon List', () => {
		var { asFragment } = render(<ListCard page="list" pokemon={ {id: '150', name: 'mewtwo', image: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/150.png'} }/>);
        expect(asFragment()).toMatchSnapshot();
    });
    
    test('List Card Content for Owned Pokemon List', () => {
		var { asFragment } = render(<ListCard page="collection" pokemon={ {id: '150', name: 'mewtwo', nickname: 'mew', image: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/150.png'} }/>);
        expect(asFragment()).toMatchSnapshot();
    });
});