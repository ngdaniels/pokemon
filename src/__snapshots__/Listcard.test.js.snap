// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Card List Card Content for Owned Pokemon List 1`] = `
<DocumentFragment>
  <table
    class="card"
    style="transform: scaleY(1);"
  >
    <tbody>
      <tr>
        <td
          class="elipsis"
          style="width: 50%;"
        >
          <span
            class="bold"
          >
            mew
          </span>
        </td>
        <td
          rowspan="2"
        >
          <img
            alt="mewtwo"
            class="pokemon-picture"
            src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/150.png"
          />
        </td>
      </tr>
      <tr>
        <td
          class="elipsis"
        >
          <div
            class="bold"
          >
            #150
            <br />
            Mewtwo
          </div>
        </td>
      </tr>
    </tbody>
    <tfoot>
      <tr>
        <td
          colspan="2"
        >
          <div
            class="flex-end"
          >
            <div
              class="card-button button-release"
              id="release-mewtwo-mew"
            >
              Release
            </div>
          </div>
        </td>
      </tr>
    </tfoot>
  </table>
</DocumentFragment>
`;

exports[`List Card List Card Content for Pokemon List 1`] = `
<DocumentFragment>
  <table
    class="card"
    style="transform: scaleY(1);"
  >
    <tbody>
      <tr>
        <td
          class="elipsis"
          style="width: 50%;"
        >
          <span
            class="bold"
          >
            #150
          </span>
        </td>
        <td
          rowspan="2"
        >
          <img
            alt="mewtwo"
            class="pokemon-picture"
            src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/150.png"
          />
        </td>
      </tr>
      <tr>
        <td
          class="elipsis"
        >
          <span
            class="bold"
          >
            Mewtwo
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
